src/foam/lib/parse/StringPStream.java

```diff
@@ -77,4 +78,12 @@ public String substring(PStream end) {
   public PStream apply(Parser ps, ParserContext x) {
     return ps.parse(this, x);
   }
+
+  public int decrement() {
+    return this.pos = this.pos == 0 ? 0 : this.pos-1;
```

Comment:
PStream is designed to be a purely functional object, its state shouldn't be mutated like this.

Instead of decrementing, you can backtrack by just remembering the previous pstream object you had a reference too.